packages/flutter/lib/src/services/binding.dart

```diff
@@ -271,66 +288,28 @@ class _DefaultBinaryMessenger extends BinaryMessenger {
     return completer.future;
   }
 
-  @override
-  Future<void> handlePlatformMessage(
-    String channel,
-    ByteData? data,
-    ui.PlatformMessageResponseCallback? callback,
-  ) async {
-    ByteData? response;
-    try {
-      final MessageHandler? handler = _handlers[channel];
-      if (handler != null) {
-        response = await handler(data);
-      } else {
-        ui.channelBuffers.push(channel, data, callback!);
-        callback = null;
-      }
-    } catch (exception, stack) {
-      FlutterError.reportError(FlutterErrorDetails(
-        exception: exception,
-        stack: stack,
-        library: 'services library',
-        context: ErrorDescription('during a platform message callback'),
-      ));
-    } finally {
-      if (callback != null) {
-        callback(response);
-      }
-    }
-  }
-
-  @override
-  Future<ByteData?>? send(String channel, ByteData? message) {
-    final MessageHandler? handler = _mockHandlers[channel];
-    if (handler != null)
-      return handler(message);
-    return _sendPlatformMessage(channel, message);
-  }
-
   @override
   void setMessageHandler(String channel, MessageHandler? handler) {
     if (handler == null) {
-      _handlers.remove(channel);
+      ui.channelBuffers.clearListener(channel);
     } else {
-      _handlers[channel] = handler;
-      ui.channelBuffers.drain(channel, (ByteData? data, ui.PlatformMessageResponseCallback callback) async {
-        await handlePlatformMessage(channel, data, callback);
+      ui.channelBuffers.setListener(channel, (ByteData? data, ui.PlatformMessageResponseCallback callback) async {
+        ByteData? response;
+        try {
+          response = await handler(data);
+        } catch (exception, stack) {
+          FlutterError.reportError(FlutterErrorDetails(
+            exception: exception,
+            stack: stack,
+            library: 'services library',
+            context: ErrorDescription('during a platform message callback'),
+          ));
+        } finally {
+          if (callback != null) {
```

Comment:
weird, why did the analyzer not catch that?
removed the `if`.

Reply:
Probably because we disable that check to allow for `assert(foo != null);` code even when foo is non-nullable:

https://github.com/flutter/flutter/blob/efbde443d2a6f810d0ceb6097fc79ed0b30bddce/analysis_options.yaml#L39-L40

Reply:
aaah, d'oh.